### ISSUE TITLE
Proposal: Faster transpilation by passing babel an AST

### ISSUE THREAD

**Issue by jaredp** [jashkenas/coffeescript#5347]
Could the new --ast mode be used to pass asts to babel for transpilation directly, instead of emitting js that gets re-parsed by babel?

`coffee -t` is slower than it ought to be, possibly 2x slower than without transpilation in my very quick test. This is a pain compared to otherwise smooth setups with nodemon, where I'm reloading on save. I'm choosing between using JSX and slowed iteration cycles vs strings of HTML and fast iteration cycles. I suspect it's the unnecessary js stringification and second parse, maybe plus composing sourcemaps.

Alternatively, CJSX syntax emitting React.createElement calls would solve my problem.

**Comment by Inve1951** [comment]
Haven't tried it myself but it should already be possible by calling the complier from your own node script.
See <https://github.com/jashkenas/coffeescript/blob/master/src/coffeescript.coffee#L115>.
```js
const CoffeeScript = require("coffeescript");
const code = _fs.readFileSync(process.argv[2], "utf8");
const ast = CoffeeScript.compile(code, { ast: true });
// const compiled = babel.transform(ast);
```

**Comment by jaredp** [comment]
I tried this via [https://repl.it/@JaredPochtar/ToughGenerousRectangles#index.js](https://repl.it/@JaredPochtar/ToughGenerousRectangles#index.js)

tl;dr: it's broken (as of today) and also slower than the alternative

**Findings:**

Performance: Passing an AST to babel appears slower (!) than converting to JS and having babel re-parse. This was pretty surprising. I haven't looked at the flamecharts, so I won't speculate why. It suggests converting from nodes->AST is as bad as nodes->JS + JS->AST, which is hard to imagine.

Correctness: Coffeescript AST output doesn't create variable declarations or compile Coffeescript `do-statements`. I'm guessing it's transforming nodes->AST before the nodes have been processed. This is probably an easy fix but a design choice for the AST flag. As it is today, passing the AST returned from Coffeescript.compile directly to Babel would not work as intended.

**Comment by GeoffreyBooth** [comment]
> Could the new `--ast` mode be used to pass asts to babel for transpilation directly, instead of emitting js that gets re-parsed by babel?

Not quite. The AST that CoffeeScript generates is its own format, not the AST defined by Babel. (And Babel’s AST is not a standard; it changes with each Babel release.) The CoffeeScript AST is _inspired_ by Babel’s, and aims to mirror it whenever possible, to enable potential use cases like you’re describing. When @helixbass was building the AST output, one idea was to have CoffeeScript generate an AST that gets passed to Prettier, for Prettier to generate JavaScript output. Such integrations would unlock lots of possibilities similar to Babel transforms, for example for a different code generator to output `const`/`let` statements instead of `var` ones, or any other differences the other generator wants to achieve.

> tl;dr: it’s broken (as of today) and also slower than the alternative

If you stick to syntax where the generated AST consists only of nodes that match one-to-one with Babel AST nodes, it _might_ work; but CoffeeScript generates some that Babel wouldn’t know what to do with, like `Existence`. Because of this fundamental incompatibility I don’t think anyone has tried it. I think the only intended recipient for the AST currently is https://github.com/helixbass/prettier-plugin-coffeescript.

> Alternatively, CJSX syntax emitting React.createElement calls would solve my problem.

TypeScript has options in its `tsconfig.json` called [`jsxFactory`](https://www.typescriptlang.org/tsconfig#jsxFactory) and [`jsxFragmentFactory`](https://www.typescriptlang.org/tsconfig#jsxFragmentFactory) which allow specifying what TypeScript should transpile JSX into (e.g. `React.createElement`). We could conceivably provide the same, and then you wouldn’t need a transpilation step; but I think there’s a strong aversion to creating more compiler options, especially options that vary the output, so I think such a proposal would likely find resistance.

Since performance is your concern, one thing you might want to consider is figuring out how to use [`esbuild`](https://github.com/evanw/esbuild) as your transpiler. It supports JSX and is much faster than Babel, so piping CoffeeScript’s output through `esbuild` might take care of the JSX conversions without a noticeable slowdown. The `--transpile` option was never meant to support _only_ Babel, so if there’s a way to integrate `esbuild` into CoffeeScript the way Babel is now, I would support such an effort; but it’s definitely easily achievable if you use some external build tool like Webpack or just pipe the output of one as the input of the other via a Node script like the examples on this thread.

**Comment by helixbass** [comment]
@GeoffreyBooth we'd discussed a "next phase" of the AST work where we tear out `.compileNode()` etc and instead add the capability to generate a "JS-transformed" Babel-compatible AST that we then feed into a simple AST -> JS printer (a la Prettier or babel/generator). And the status of that was that (since my initial AST work was more oriented around this than a "pure" Coffeescript AST) at one point I had this almost entirely working (I think comment printing was the only major TODO)

So that would presumably be aligned with this type of ask since then I think it legitimately would work to pass our "JS-transformed" AST directly to eg Babel rather than printing/re-parsing first

**Comment by jaredp** [comment]
Very surprisingly, esbuild seems to be the slowest option of them all.

One run gives me (in milliseconds, min of 100 runs after a 100 run warmup)
```
via `-t` (JS->babel): 13.872173000127077
via AST->babel: 14.50042099878192
without transpilation: 4.3442809991538525
via esbuild: 20.119036998599768
```

I'm now confident my measurement's wrong?

**Comment by edemaine** [comment]
> I think the only intended recipient for the AST currently is https://github.com/helixbass/prettier-plugin-coffeescript.

FWIW, another recipient is https://github.com/helixbass/eslint-plugin-coffee for better `eslint` processing of CoffeeScript. I was about to suggest linking to these tools from coffeescript.org, but I see both are already mentioned in https://coffeescript.org/#changelog 🙂

**Comment by GeoffreyBooth** [comment]
> I was about to suggest linking to these tools from coffeescript.org

There was a PR to add a new “Integrations” section to the docs to make these more prominent. #5334. We should really finish that 😄
